crates/cubecl-wgpu/src/runtime.rs

```diff
@@ -90,95 +90,92 @@ impl Default for RuntimeOptions {
-    RUNTIME.register(device, client)
+    future::block_on(init_async::<G>(device, options))
 }
 
-pub async fn create_wgpu_setup<G: GraphicsApi, C: WgpuCompiler>(
-    device: &WgpuDevice,
-) -> (Arc<wgpu::Adapter>, Arc<wgpu::Device>, Arc<wgpu::Queue>) {
-    let (device_wgpu, queue, adapter) = select_device::<G, C>(device).await;
-
-    log::info!(
-        "Created wgpu compute server on device {:?} => {:?}",
-        device,
-        adapter.get_info()
-    );
-    (Arc::new(adapter), Arc::new(device_wgpu), Arc::new(queue))
+/// Like [`init_sync`], but async, necessary for wasm.
+pub async fn init_async<G: GraphicsApi>(device: &WgpuDevice, options: RuntimeOptions) -> WgpuSetup {
+    let setup = create_wgpu_setup::<G, WgslCompiler>(device).await;
+    let return_setup = setup.clone();
+    let client = create_client(setup, options);
+    RUNTIME.register(device, client);
+    return_setup
 }
 
 pub fn create_client<C: WgpuCompiler>(
-    adapter: Arc<wgpu::Adapter>,
-    device_wgpu: Arc<wgpu::Device>,
-    queue: Arc<wgpu::Queue>,
+    setup: WgpuSetup,
     options: RuntimeOptions,
 ) -> ComputeClient<WgpuServer<C>, MutexComputeChannel<WgpuServer<C>>> {
-    let limits = device_wgpu.limits();
+    let limits = setup.device.limits();
     let mem_props = MemoryDeviceProperties {
         max_page_size: limits.max_storage_buffer_binding_size as u64,
         alignment: WgpuStorage::ALIGNMENT.max(limits.min_storage_buffer_offset_alignment as u64),
     };
 
     let memory_management = init_memory_management(
-        device_wgpu.clone(),
+        setup.device.clone(),
         mem_props.clone(),
         options.memory_config,
     );
     let server = WgpuServer::new(
         memory_management,
-        device_wgpu.clone(),
-        queue,
+        setup.device.clone(),
+        setup.queue,
         options.tasks_max,
     );
     let channel = MutexComputeChannel::new(server);
 
-    let features = adapter.features();
+    let features = setup.adapter.features();
     let mut device_props = DeviceProperties::new(&[], mem_props);
     if features.contains(wgpu::Features::SUBGROUP) {
         device_props.register_feature(Feature::Subcube);
     }
-    C::register_features(&adapter, &device_wgpu, &mut device_props);
+    C::register_features(&setup.adapter, &setup.device, &mut device_props);
     ComputeClient::new(channel, device_props)
 }
 
 /// Select the wgpu device and queue based on the provided [device](WgpuDevice).
-pub async fn select_device<G: GraphicsApi, C: WgpuCompiler>(
-    device: &WgpuDevice,
-) -> (wgpu::Device, wgpu::Queue, wgpu::Adapter) {
-    #[cfg(target_family = "wasm")]
-    let adapter = select_adapter::<G>(device).await;
-
-    #[cfg(not(target_family = "wasm"))]
-    let adapter = select_adapter::<G>(device);
-
+pub async fn create_wgpu_setup<G: GraphicsApi, C: WgpuCompiler>(device: &WgpuDevice) -> WgpuSetup {
+    let (instance, adapter) = select_adapter::<G>(device).await;
```

Comment:
Hmm yeah it's confusing with the functions `init_async` and `init_sync`. Maybe instead of `init` we could have `register_setup_sync` and `register_setup_async`. Also we could remove `wgpu` from the function name here, since it's in the wgpu project: `create_setup` is probably sufficient.

Reply:
Yeah this could do with better naming. Register didn't feel right. This function does create new things, not just sign them up somewhere. Initialization is better imo (or maybe create). I went for this in the end:

- init_sync -> init_device_sync.
- init -> init_device
- init_existing_device -> init_device_on_setup
- create_client -> create_client_on_setup, and is now private, doesn't need to be public
- create_wgpu_setup -> create_setup_for_device, and is now private, doesn't need to be public

Lmk how that looks.